jupyterhub/tests/selenium/test_browser.py

```diff
@@ -0,0 +1,419 @@
+            )
+        assert next_url.endswith(f"/user/{user}/")
+        print("Current Link 2: " + browser.current_url)
+
+
+"""TBD"""
+
+
+@pytest.mark.parametrize(
+    'running, next_url, location, params',
+    [
+        # default URL if next not specified, for both running and not
+        (True, '', '', None),
+        (False, '', '', None),
+        # next_url is respected
+        (False, '/hub/admin', '/hub/admin', None),
+        (False, '/user/other', '/hub/user/other', None),
+        (False, '/absolute', '/absolute', None),
+        (False, '/has?query#andhash', '/has?query#andhash', None),
+        # :// in query string or fragment
+        (False, '/has?repo=https/host.git', '/has?repo=https/host.git', None),
+        (False, '/has?repo=https://host.git', '/has?repo=https://host.git', None),
+        (False, '/has#repo=https://host.git', '/has#repo=https://host.git', None),
+        # next_url outside is not allowed
+        (False, 'relative/path', '', None),
+        (False, 'https://other.domain', '', None),
+        (False, 'ftp://other.domain', '', None),
+        (False, '//other.domain', '', None),
+        (False, '///other.domain/triple', '', None),
+        (False, '\\\\other.domain/backslashes', '', None),
+        # params are handled correctly (ignored if ?next= specified)
+        (
+            True,
+            '/hub/admin?left=1&right=2',
+            'hub/admin?left=1&right=2',
+            {"left": "abc"},
+        ),
+        (False, '/hub/admin', 'hub/admin', [('left', 1), ('right', 2)]),
+        (True, '', '', {"keep": "yes"}),
+        (False, '', '', {"keep": "yes"}),
+    ],
+)
+async def test_login_redirect(app, browser, running, next_url, location, params, user):
+
+    if location:
+        location = ujoin(app.base_url, location)
+        print("if location is:" + location)
+    elif running:
+        # location not specified,
+        location = user.url
+        if params:
+            location = url_concat(location, params)
+    else:
+        # use default url
+        location = ujoin(app.base_url, 'hub/spawn')
+        if params:
+            location = url_concat(location, params)
+
+    url = 'login'
+    if params:
+        url = url_concat(url, params)
+    if next_url:
+        if next_url.startswith('/') and not (
+            next_url.startswith("//") or urlparse(next_url).netloc
+        ):
+            next_url = ujoin(app.base_url, next_url, '')
+        url = url_concat(url, dict(next=next_url))
+
+    if running and not user.active:
+        # ensure running
+        await user.spawn()
+    elif user.active and not running:
+        # ensure not running
+        await user.stop()
+    time.sleep(5)
+    # open_url(app,browser)
+    await in_thread(browser.get, (ujoin(public_url(app), url)))
+    await login(browser, user='test_user', pass_w='test_user')
+
+    # await get_page(url, app, allow_redirects=False)
+    time.sleep(50)
```

Comment:
Make sure to remove all `time.sleep`, which are blocking and prevent processing on the server. We _may_ need a few short `await asyncio.sleep`, but let's try to wait for specific conditions as much as possible.